display_list/geometry/dl_region.cc

```diff
@@ -528,16 +521,23 @@ DlRegion DlRegion::MakeIntersection(const DlRegion& a, const DlRegion& b) {
   auto b_it = b.lines_.begin();
   auto b_end = b.lines_.end();
 
-  // When intersecting single rectangle with a complex region use binary
-  // search to find the first line that intersects the rectangle.
-  if (b_it == b_end - 1) {
-    a_it = std::upper_bound(
-        a_it, a_end, b_it->top,
-        [](int32_t top, const SpanLine& line) { return top < line.bottom; });
-  } else if (a_it == a_end - 1) {
-    b_it = std::upper_bound(
-        b_it, b_end, a_it->top,
-        [](int32_t top, const SpanLine& line) { return top < line.bottom; });
+  FML_DCHECK(a_it != a_end && b_it != b_end);
+
+  auto a_len = a_end - a_it;
+  auto b_len = b_end - b_it;
+
+  if (a_len > kBinarySearchThreshold &&
+      a_it[kBinarySearchThreshold].bottom <= b_it->top) {
+    a_it = std::lower_bound(
+        a.lines_.begin() + kBinarySearchThreshold + 1, a.lines_.end(),
+        b_it->top,
+        [](const SpanLine& line, int32_t top) { return line.bottom <= top; });
+  } else if (b_len > kBinarySearchThreshold &&
```

Comment:
Would it be better to check them both with 2 separate `if`s rather than using `else if`? (Or maybe not, see next comment)

Reply:
Ah, maybe not. It looks like if the first one triggers then `a` has some "catching up to do" and so the second one would not trigger because `b` obviously started later than `a`. So, `else if` is the optimal logic. Am I reading that right?

Reply:
Yes. That's the idea.